File: Frontend/src/types/incident.ts
// =============================================================================
// Incident Type Definitions — maps to backend C# DTOs & Enums
// =============================================================================

import type { ApiResult } from "@/types/agent";
export type { ApiResult };

// ---------------------------------------------------------------------------
// Enums (string literal unions mirroring backend C# enums)
// ---------------------------------------------------------------------------

export const INCIDENT_STATUSES = [
  "Open",
  "Investigating",
  "Mitigated",
  "Resolved",
  "Closed",
  "Escalated",
] as const;
export type IncidentStatus = (typeof INCIDENT_STATUSES)[number];

export const INCIDENT_SEVERITIES = [
  "P1",
  "P2",
  "P3",
  "P4",
] as const;
export type IncidentSeverity = (typeof INCIDENT_SEVERITIES)[number];

export const INCIDENT_ROUTES = [
  "SopExecution",
  "RootCauseAnalysis",
  "SopGeneration",
  "FallbackRca",
] as const;
export type IncidentRoute = (typeof INCIDENT_ROUTES)[number];

export const TIMELINE_EVENT_TYPES = [
  "AlertReceived",
  "StatusChanged",
  "AgentMessage",
  "ToolInvoked",
  "ToolCall",
  "SopStepCompleted",
  "RcaCompleted",
  "SopGenerated",
  "Resolved",
  "Escalated",
  "ManualNote",
  "Timeout",
  "HumanIntervention",
  "InterventionRequested",
  "ToolApprovalRequested",
  "ToolApprovalResponded",
  "SopFallbackToRca",
  "SopAutoDisabled",
] as const;
export type TimelineEventType = (typeof TIMELINE_EVENT_TYPES)[number];

// ---------------------------------------------------------------------------
// Display labels
// ---------------------------------------------------------------------------

export const SEVERITY_LABELS: Record<IncidentSeverity, string> = {
  P1: "P1 — Critical",
  P2: "P2 — High",
  P3: "P3 — Medium",
  P4: "P4 — Low",
};

export const STATUS_LABELS: Record<IncidentStatus, string> = {
  Open: "待处理",
  Investigating: "处理中",
  Mitigated: "已缓解",
  Resolved: "已解决",
  Closed: "已关闭",
  Escalated: "已上报",
};

export const ROUTE_LABELS: Record<IncidentRoute, string> = {
  SopExecution: "SOP 自动执行",
  RootCauseAnalysis: "根因分析",
  SopGeneration: "SOP 生成",
  FallbackRca: "降级根因分析",
};

// ---------------------------------------------------------------------------
// DTOs
// ---------------------------------------------------------------------------

/** Maps to backend IncidentSummaryDto */
export interface IncidentSummary {
  id: string;
  title: string;
  status: IncidentStatus;
  severity: IncidentSeverity;
  route: IncidentRoute;
  alertName: string;
  alertFingerprint: string | null;
  alertRuleId: string;
  createdAt: string;
  updatedAt: string | null;
}

// ---------------------------------------------------------------------------
// Enums — Evaluation / Step Execution
// ---------------------------------------------------------------------------

export type RcaAccuracyRating = "Accurate" | "PartiallyAccurate" | "Inaccurate" | "NotApplicable";
export type SopEffectivenessRating = "Effective" | "PartiallyEffective" | "Ineffective";
export type SopStepType = "Structured" | "Freeform";
export type StepExecutionStatus = "Pending" | "Running" | "Completed" | "Failed" | "Skipped";

// ---------------------------------------------------------------------------
// Post-Mortem Annotation (Spec 023)
// ---------------------------------------------------------------------------

export interface PostMortemAnnotation {
  actualRootCause: string;
  rcaAccuracy: RcaAccuracyRating;
  sopEffectiveness: SopEffectivenessRating | null;
  improvementNotes: string | null;
  annotatedBy: string;
  annotatedAt: string;
}

// ---------------------------------------------------------------------------
// SOP Step Execution (Spec 024)
// ---------------------------------------------------------------------------

export interface SopStepDefinition {
  stepNumber: number;
  description: string;
  stepType: SopStepType;
  toolName: string | null;
  parameterTemplate: unknown | null;
  expectedOutcome: string | null;
  timeoutSeconds: number;
  requiresApproval: boolean;
}

export interface SopStepExecution {
  stepNumber: number;
  status: StepExecutionStatus;
  startedAt: string | null;
  completedAt: string | null;
  toolCallResult: unknown | null;
  agentJudgment: string | null;
  retryCount: number;
  parameterDeviations: string[] | null;
  errorMessage: string | null;
}

/** Maps to backend IncidentDetailDto */
export interface IncidentDetail {
  id: string;
  title: string;
  status: IncidentStatus;
  severity: IncidentSeverity;
  route: IncidentRoute;
  alertName: string;
  alertFingerprint: string | null;
  alertRuleId: string;
  conversationId: string | null;
  rootCause: string | null;
  resolution: string | null;
  generatedSopId: string | null;
  alertLabels: Record<string, string> | null;
  timeToDetect: string | null;
  timeToResolve: string | null;
  timeline: IncidentTimelineItem[];
  createdAt: string;
  updatedAt: string | null;

  // Spec 023 — Post-mortem
  postMortem: PostMortemAnnotation | null;

  // Spec 024 — Step execution
  sopSteps: SopStepDefinition[] | null;
  stepExecutions: SopStepExecution[];

  // Spec 025 — Fallback
  fallbackFrom: string | null;
  fallbackReason: string | null;
}

/** Maps to backend IncidentTimelineItemDto */
export interface IncidentTimelineItem {
  eventType: TimelineEventType;
  summary: string;
  timestamp: string;
  actorAgentId: string | null;
  metadata: Record<string, string> | null;
}

// ---------------------------------------------------------------------------
// SignalR Events (mirrors backend IIncidentClient)
// ---------------------------------------------------------------------------

export interface IncidentCreatedEvent {
  incidentId: string;
  title: string;
  status: string;
  severity: string;
  route: string;
  alertName: string;
  alertRuleId: string;
  createdAt: string;
}

export interface IncidentStatusChangedEvent {
  incidentId: string;
  oldStatus: string;
  newStatus: string;
  timestamp: string;
}

export interface TimelineEventAddedPayload {
  incidentId: string;
  eventType: string;
  summary: string;
  timestamp: string;
  actorAgentId: string | null;
  metadata: Record<string, string> | null;
}

export interface ChatMessagePayload {
  incidentId: string;
  role: string;
  content: string;
  agentName: string | null;
  timestamp: string | null;
}

export interface IncidentResolvedEvent {
  incidentId: string;
  resolution: string | null;
  resolvedAt: string;
}

export interface RcaCompletedEvent {
  incidentId: string;
  rootCause: string;
  timestamp: string;
}

export interface SopGeneratedEvent {
  incidentId: string;
  skillId: string;
  sopName: string;
  timestamp: string;
}

export interface AgentProcessingChangedEvent {
  incidentId: string;
  isProcessing: boolean;
  agentName: string | null;
  timestamp: string;
}

export interface HumanInterventionAcknowledgedEvent {
  incidentId: string;
  timestamp: string;
}

export interface IncidentTimeoutEvent {
  incidentId: string;
  reason: string;
  timestamp: string;
}

export interface IncidentEscalatedEvent {
  incidentId: string;
  reason: string;
  timestamp: string;
}

// ---------------------------------------------------------------------------
// Structured Intervention (Feature A/B/C)
// ---------------------------------------------------------------------------

export type InterventionRequestType = "ToolApproval" | "FreeTextInput" | "Choice";
export type InterventionResponseType = "Approved" | "Rejected" | "TextInput" | "ChoiceSelected";

export interface InterventionRequestPayload {
  incidentId: string;
  requestId: string;
  type: InterventionRequestType;
  prompt: string;
  createdAt: string;
  toolName?: string | null;
  toolCallId?: string | null;
  toolArguments?: Record<string, unknown> | null;
  choices?: string[] | null;
}

export interface InterventionRequestResolvedPayload {
  incidentId: string;
  requestId: string;
  responseType: string;
  responseContent?: string | null;
  approved?: boolean | null;
  operatorName?: string | null;
  timestamp?: string | null;
}

export interface InterventionResponseBody {
  responseType: InterventionResponseType;
  content?: string;
  approved?: boolean;
  operatorName?: string;
}

File: Backend/CoreSRE.Domain/Entities/Incident.cs
using System.Text.Json;
using CoreSRE.Domain.Enums;
using CoreSRE.Domain.ValueObjects;

namespace CoreSRE.Domain.Entities;

/// <summary>
/// 故障事件（Incident）。由告警触发，记录完整处置生命周期。
/// </summary>
public class Incident : BaseEntity
{
    /// <summary>事故标题（自动生成或从告警 annotations.summary 提取）</summary>
    public string Title { get; private set; } = string.Empty;

    /// <summary>事故严重等级</summary>
    public IncidentSeverity Severity { get; private set; } = IncidentSeverity.P3;

    /// <summary>事故状态</summary>
    public IncidentStatus Status { get; private set; } = IncidentStatus.Open;

    /// <summary>触发的路由规则 ID</summary>
    public Guid? AlertRuleId { get; private set; }

    /// <summary>Alertmanager fingerprint（去重键）</summary>
    public string AlertFingerprint { get; private set; } = string.Empty;

    /// <summary>原始告警 JSON（审计用）</summary>
    public JsonDocument? AlertPayload { get; private set; }

    /// <summary>告警标签快照</summary>
    public Dictionary<string, string> AlertLabels { get; private set; } = new();

    /// <summary>处置链路</summary>
    public IncidentRoute Route { get; private set; }

    /// <summary>Agent 对话 ID（链接到 Conversation 实体）</summary>
    public Guid? ConversationId { get; private set; }

    /// <summary>使用的 SOP（SkillRegistration ID）</summary>
    public Guid? SopId { get; private set; }

    /// <summary>根因分析结论</summary>
    public string? RootCause { get; private set; }

    /// <summary>处置结论</summary>
    public string? Resolution { get; private set; }

    /// <summary>根因链路生成的新 SOP（如果有）</summary>
    public Guid? GeneratedSopId { get; private set; }

    /// <summary>开始处理时间</summary>
    public DateTime StartedAt { get; private set; } = DateTime.UtcNow;

    /// <summary>解决时间</summary>
    public DateTime? ResolvedAt { get; private set; }

    /// <summary>MTTD — 告警触发到开始处理（毫秒）</summary>
    public long? TimeToDetectMs { get; private set; }

    /// <summary>MTTR — 开始处理到解决（毫秒）</summary>
    public long? TimeToResolveMs { get; private set; }

    /// <summary>事件时间线</summary>
    public List<IncidentTimelineVO> Timeline { get; private set; } = [];

    private Incident() { } // EF Core

    /// <summary>创建 SOP 执行链路的 Incident</summary>
    public static Incident CreateForSopExecution(
        string title,
        IncidentSeverity severity,
        Guid alertRuleId,
        string alertFingerprint,
        JsonDocument? alertPayload,
        Dictionary<string, string> alertLabels,
        Guid sopId)
    {
        var incident = new Incident
        {
            Title = title,
            Severity = severity,
            Status = IncidentStatus.Open,
            AlertRuleId = alertRuleId,
            AlertFingerprint = alertFingerprint,
            AlertPayload = alertPayload,
            AlertLabels = alertLabels,
            Route = IncidentRoute.SopExecution,
            SopId = sopId,
            StartedAt = DateTime.UtcNow
        };
        incident.AddTimelineEvent(TimelineEventType.AlertReceived, $"告警已接收: {title}");
        return incident;
    }

    /// <summary>创建根因分析链路的 Incident</summary>
    public static Incident CreateForRootCauseAnalysis(
        string title,
        IncidentSeverity severity,
        Guid alertRuleId,
        string alertFingerprint,
        JsonDocument? alertPayload,
        Dictionary<string, string> alertLabels)
    {
        var incident = new Incident
        {
            Title = title,
            Severity = severity,
            Status = IncidentStatus.Open,
            AlertRuleId = alertRuleId,
            AlertFingerprint = alertFingerprint,
            AlertPayload = alertPayload,
            AlertLabels = alertLabels,
            Route = IncidentRoute.RootCauseAnalysis,
            StartedAt = DateTime.UtcNow
        };
        incident.AddTimelineEvent(TimelineEventType.AlertReceived, $"告警已接收: {title}");
        return incident;
    }

    /// <summary>追加时间线事件</summary>
    public void AddTimelineEvent(TimelineEventType eventType, string summary, string? details = null)
    {
        Timeline.Add(IncidentTimelineVO.Create(eventType, summary, details));
    }

    /// <summary>追加时间线事件（完整 VO）</summary>
    public void AddTimelineEvent(IncidentTimelineVO timelineEvent)
    {
        Timeline.Add(timelineEvent);
    }

    /// <summary>关联对话</summary>
    public void SetConversation(Guid conversationId)
    {
        ConversationId = conversationId;
    }

    /// <summary>更新状态（含生命周期校验）</summary>
    public void TransitionTo(IncidentStatus newStatus)
    {
        if (!IsValidTransition(Status, newStatus))
            throw new InvalidOperationException(
                $"Invalid incident status transition: {Status} → {newStatus}");

        var oldStatus = Status;
        Status = newStatus;

        if (newStatus == IncidentStatus.Resolved || newStatus == IncidentStatus.Closed)
        {
            ResolvedAt = DateTime.UtcNow;
            TimeToResolveMs = (long)(ResolvedAt.Value - StartedAt).TotalMilliseconds;
        }

        AddTimelineEvent(TimelineEventType.StatusChanged,
            $"状态变更: {oldStatus} → {newStatus}");
    }

    /// <summary>记录根因分析结论</summary>
    public void SetRootCause(string rootCause)
    {
        RootCause = rootCause;
        AddTimelineEvent(TimelineEventType.RootCauseFound, "根因分析完成", rootCause);
    }

    /// <summary>记录处置结论并标记已解决</summary>
    public void Resolve(string resolution)
    {
        Resolution = resolution;
        TransitionTo(IncidentStatus.Resolved);
        AddTimelineEvent(TimelineEventType.Resolved, "事故已解决", resolution);
    }

    /// <summary>记录生成的 SOP</summary>
    public void SetGeneratedSop(Guid sopId)
    {
        GeneratedSopId = sopId;
        AddTimelineEvent(TimelineEventType.SopGenerated, $"已自动生成 SOP: {sopId}");
    }

    /// <summary>设置 MTTD</summary>
    public void SetTimeToDetect(DateTime alertFiredAt)
    {
        TimeToDetectMs = (long)(StartedAt - alertFiredAt).TotalMilliseconds;
    }

    /// <summary>上报（人工介入）</summary>
    public void Escalate(string reason)
    {
        AddTimelineEvent(TimelineEventType.Escalated, $"已上报: {reason}");
    }

    /// <summary>状态流转合法性校验</summary>
    private static bool IsValidTransition(IncidentStatus from, IncidentStatus to)
    {
        return (from, to) switch
        {
            (IncidentStatus.Open, IncidentStatus.Investigating) => true,
            (IncidentStatus.Open, IncidentStatus.Resolved) => true,       // 快速解决
            (IncidentStatus.Open, IncidentStatus.Closed) => true,         // 误报关闭
            (IncidentStatus.Investigating, IncidentStatus.Mitigated) => true,
            (IncidentStatus.Investigating, IncidentStatus.Resolved) => true,
            (IncidentStatus.Investigating, IncidentStatus.Closed) => true, // 取消
            (IncidentStatus.Mitigated, IncidentStatus.Resolved) => true,
            (IncidentStatus.Mitigated, IncidentStatus.Investigating) => true, // 回退
            (IncidentStatus.Resolved, IncidentStatus.Closed) => true,
            (IncidentStatus.Resolved, IncidentStatus.Investigating) => true,  // 重开
            _ => false
        };
    }
}
